### PR TITLE
feat: implement tracking as per spec

### PR DESCRIPTION
**Pull request by toddbaert**
:mega: This was a draft for a while, but is now ready for review! :mega: 

This implements tracking as per spec, in the server, web, and react SDKs.
I don't think the Angular or Nest SDKs need specific implementations, but please advise (cc @luizgribeiro @lukas-reining).

Fixes: https://github.com/open-feature/js-sdk/issues/1033
Fixes: https://github.com/open-feature/js-sdk/issues/1034

### CODE REVIEW

File: packages/shared/src/tracking/occurrence.ts
Comment (toddbaert) [inline packages/shared/src/tracking/occurrence.ts:18]:
```suggestion
  /**
   * A numeric value associated with this event.
   */
```

File: packages/web/src/tracking/tracking.ts
Comment (beeme1mr) [inline packages/web/src/tracking/tracking.ts:1]:
```suggestion
import types { OccurrenceDetails } from '@openfeature/core';
```

File: packages/web/src/provider/provider.ts
Comment (moredip) [inline packages/web/src/provider/provider.ts:81]: I think `event` or `eventName` is more idiomatic and recognizable than `occurenceKey` (closer to the defacto standard of [Segment's SDK](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#track)) 
```suggestion
    track?(event: string, context: EvaluationContext, occurrenceDetails: OccurrenceDetails): void;
```

File: packages/web/src/tracking/tracking.ts
Comment (weyert) [reply packages/web/src/tracking/tracking.ts:1]: `import type`

File: packages/server/src/provider/provider.ts
Comment (lukas-reining) [inline packages/server/src/provider/provider.ts:73]: Are we doing these later?

File: packages/shared/src/tracking/occurrence.ts
Comment (lukas-reining) [inline packages/shared/src/tracking/occurrence.ts:12]: This is copy-paste right? Seems to make for sense for evaluation context.

Comment (toddbaert) [reply packages/shared/src/tracking/occurrence.ts:12]: Yep - this is still quite rough - we aren't settle on terminology yet and comments aren't updated.

File: packages/server/src/provider/provider.ts
Comment (toddbaert) [reply packages/server/src/provider/provider.ts:73]: https://github.com/open-feature/js-sdk/pull/1020#discussion_r1779083185

File: packages/shared/src/tracking/occurrence.ts
Comment (lukas-reining) [reply packages/shared/src/tracking/occurrence.ts:12]: Sure this is great, just wanted to be sure!

**Review by aepfli — APPROVED** [review]

File: packages/web/src/tracking/tracking.ts
Comment (toddbaert) [reply packages/web/src/tracking/tracking.ts]: I've fixed this is a separate PR with a new eslint rule, nice catch.

**Comment by toddbaert** [comment]
This is ready for review now.

File: packages/web/src/client/internal/open-feature-client.ts
Comment (beeme1mr) [inline packages/web/src/client/internal/open-feature-client.ts:187]: Do we want to throw if the provider is unhealthy? I expected the track method to never throw.

File: packages/web/src/client/internal/open-feature-client.ts
Comment (beeme1mr) [inline packages/web/src/client/internal/open-feature-client.ts:189]: It may be worth adding a comment explaining why you're creating a new context object. I'm assuming it's because you don't want the provider to be able to modify context in the track method.

File: packages/server/src/client/internal/open-feature-client.ts
Comment (beeme1mr) [inline packages/server/src/client/internal/open-feature-client.ts:234]: Shouldn't we log and swallow any exceptions thrown by the provider?

Comment (toddbaert) [reply packages/server/src/client/internal/open-feature-client.ts:234]: Definitely; this would fall into our spec rules for client ops not abnormally terminating. I will add some error handling.

File: packages/web/src/client/internal/open-feature-client.ts
Comment (toddbaert) [reply packages/web/src/client/internal/open-feature-client.ts:187]: We want to throw here but catch it, I think. I just made it short circuit like all the other methods.

File: packages/server/src/client/internal/open-feature-client.ts
Comment (toddbaert) [reply packages/server/src/client/internal/open-feature-client.ts]: Added

File: packages/web/src/client/internal/open-feature-client.ts
Comment (toddbaert) [reply packages/web/src/client/internal/open-feature-client.ts]: I did this and also froze the context.

File: packages/web/src/client/internal/open-feature-client.ts
Comment (toddbaert) [reply packages/web/src/client/internal/open-feature-client.ts]: done

File: packages/server/src/client/internal/open-feature-client.ts
Comment (toddbaert) [reply packages/server/src/client/internal/open-feature-client.ts:401]: This is harder than it sounds because the server/web actually both use different enums since there's more types in client.

File: packages/react/src/tracking/use-track.ts
Comment (beeme1mr) [inline packages/react/src/tracking/use-track.ts:5]: Couldn't this type live in core?

File: packages/web/src/client/internal/open-feature-client.ts
Comment (beeme1mr) [inline packages/web/src/client/internal/open-feature-client.ts:193]: Please move this under the if condition.

File: packages/web/src/client/internal/open-feature-client.ts
Comment (beeme1mr) [inline packages/web/src/client/internal/open-feature-client.ts:337]: Perhaps we could move this to a utils method in core to avoid duplication.

File: packages/react/src/tracking/use-track.ts
Comment (toddbaert) [reply packages/react/src/tracking/use-track.ts:5]: It's used only in React; it's the type of the exported track hook - it happens to look like the track method but it's not a method.

File: packages/web/src/client/internal/open-feature-client.ts
Comment (toddbaert) [reply packages/web/src/client/internal/open-feature-client.ts:337]: See: https://github.com/open-feature/js-sdk/pull/1020#discussion_r1804558887

File: packages/react/src/tracking/use-track.ts
Comment (toddbaert) [reply packages/react/src/tracking/use-track.ts:5]: I can do `track: Tracking['track'];` actually, since this function actually matches that type.

**Review by beeme1mr — APPROVED** [review]

File: packages/web/src/provider/in-memory-provider/in-memory-provider.ts
Comment (wichopy) [inline packages/web/src/provider/in-memory-provider/in-memory-provider.ts:41]: Since you are trying to merge old and new configs, shouldn't the new config come last?
```suggestion
      const flagsChanged = Object.entries({...this._flagConfiguration,...flagConfiguration })
```

Comment (wichopy) [reply packages/web/src/provider/in-memory-provider/in-memory-provider.ts:41]: maybe need a test case for this?

Comment (toddbaert) [reply packages/web/src/provider/in-memory-provider/in-memory-provider.ts:41]: reverted

**Review by wichopy — APPROVED** [review]
Lgtm. Would be nice to hook it up to @beeme1mr s test app?

**Review by aepfli — APPROVED** [review]

**Comment by toddbaert** [comment]
> Lgtm. Would be nice to hook it up to @beeme1mr s test app?

That's the plan!

File: packages/server/src/client/internal/open-feature-client.ts
Comment (lukas-reining) [inline packages/server/src/client/internal/open-feature-client.ts:394]: Maybe instead if using the OpenFeature global API to get the context we could add a context accessor function as we do here: https://github.com/open-feature/js-sdk/pull/1020/files#diff-8a151eb52e8ba25881ad3b837c98a32807dabf7b7f771518248edb6206e5add9R51
I would like this as we would not have a circular dependency between the global API and the client then.

But this would also apply to `OpenFeature.getHooks` used in this class and it might be out of context for this API. So we can resolve this, just wanted to mention it and see if we want to do it now.

File: packages/server/src/provider/provider.ts
Comment (lukas-reining) [inline packages/server/src/provider/provider.ts:76]: This is the same for server and client, I think this could live in the `CommonProvider` like `initialize` and `onClose`.

**Review by lukas-reining — APPROVED** [review]
Looks good! Left two comments :)

File: packages/server/src/provider/provider.ts
Comment (toddbaert) [reply packages/server/src/provider/provider.ts:76]: Oh you are right I think. Will do.

**Comment by toddbaert** [comment]
I'm merge this tomorrow unless I hear objections

File: packages/server/src/client/internal/open-feature-client.ts
Comment (lukas-reining) [reply packages/server/src/client/internal/open-feature-client.ts]: Great just a nice thing to "clean up"! :)

**Pull request merged**
